processor/src/main/java/com/linecorp/decaton/processor/runtime/internal/Utils.java

```diff
@@ -202,4 +207,32 @@ public static CompletableFuture<Void> runInParallel(String subject, Collection<T
         return CompletableFuture.allOf(results)
                                 .whenComplete((v, e) -> executor.shutdown());
     }
+
+    private static final ScheduledExecutorService scheduledExecutor;
+    static {
+        ((ScheduledThreadPoolExecutor) (scheduledExecutor =
+                Executors.newScheduledThreadPool(1, r -> {
+                    Thread th = namedThreadFactory(
+                            "CompletableFutureCanceller").newThread(r);
+                    th.setDaemon(true);
+                    return th;
+                }))).setRemoveOnCancelPolicy(true);
+    }
+
+    public static <T> CompletableFuture<T> completeOnTimeout(CompletableFuture<T> cf, T value, long timeoutMs) {
+        if (cf.isDone()) {
+            return cf;
+        }
+        ScheduledFuture<?> cancelFut = scheduledExecutor.schedule(() -> {
+            if (!cf.isDone()) {
+                cf.complete(value);
```

Comment:
Since this completion handler may include `destroyThreadProcessor` call which is potentially costly, shouldn't we have more pool size than `1` ? (like available processor num ?)

Reply:
IIUC, I think the first argument is "corePoolSize", so my understanding of the behavior is, when there's a running scheduled task at the time of next scheduled execution occurs, the executor creates a new thread for processing it (and it's count is unbound) so a scheduled task execution never blocks the others.

Reply:
hm, I think your concern is correct. fixing.